File: 1488-avoid-flood-in-the-city/1488-avoid-flood-in-the-city.cpp
class Solution {
public:
    vector<int> avoidFlood(vector<int>& rains) {
        int n = rains.size();
        vector<int> ans(n, 1);
        unordered_map<int, int> lake_day; // lake -> 上次下雨的日子
        set<int> dry_days; // 可用乾旱的日子（升序）

        for (int i = 0; i < n; ++i){
            int lake = rains[i];
            if (lake > 0){
                ans[i] = -1;
                // 如果這個湖之前下過雨
                if (lake_day.count(lake)){
                    int prev_day = lake_day[lake];
                    // 找出第一個比 prev_day 大的乾旱日
                    auto it = dry_days.upper_bound(prev_day);
                    if (it == dry_days.end()){
                        return {}; // 無法避免洪水
                    }
                    int dry_day = *it;
                    ans[dry_day] = lake;    // 乾這個湖
                    dry_days.erase(it); // 從可用乾旱日中移除
                }
                lake_day[lake] = i;
            } else{
                // 沒下雨，這天可用來乾湖
                dry_days.insert(i);
            }
        }
        return ans;
    }
};

File: 1488-avoid-flood-in-the-city/1488-avoid-flood-in-the-city.py
class Solution:
    def avoidFlood(self, rains: List[int]) -> List[int]:
        lake_day = {} # lake -> 上次下雨的日子
        dry_days = [] # 可用乾旱日（升序
        n = len(rains)
        ans = [-1] * n

        for i, lake in enumerate(rains):
            if lake > 0:
                # 如果湖已滿，需要找出下一個可乾的日子
                if lake in lake_day:
                    prev_day = lake_day[lake]
                    # 找出在 prev_day 之後的乾旱日
                    idx = bisect_right(dry_days, prev_day)
                    if idx == len(dry_days):
                        return [] # 沒有可以乾的日子 → 洪水
                    dry_day = dry_days.pop(idx)
                    ans[dry_day] = lake # 乾這個湖
                lake_day[lake] = i
            else:
                # 可乾旱的日子
                insort(dry_days, i)
                ans[i] = 1 # 先填預設值
        
        return ans